File: src/leetcode/python/AddStrings_415.py
class AddStrings_415(object):
    def addStrings(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        if len(num2) > len(num1):
            num1, num2 = num2, num1
        res = []

        def addStr(l, r, c):
            t = int(l) + int(r)
            if c:
                t += 1
            res.append(str(t % 10))
            return t >= 10

        nl, nr = len(num1) - 1, len(num2) - 1
        b = False
        while nr >= 0:
            b = addStr(num1[nl], num2[nr], b)
            nl, nr = nl - 1, nr - 1
        while nl >= 0:
            b = addStr(num1[nl], '0', b)
            nl -= 1
        if b:
            res.append("1")
        return ''.join(res[::-1])

File: src/leetcode/java/AddStrings_415.java
package leetcode.java;

public class AddStrings_415 {
  public String addStrings(String num1, String num2) {
    StringBuilder sum = new StringBuilder();
    char[] left = num1.length() >= num2.length() ? num1.toCharArray() : num2.toCharArray();
    char[] right = num1.length() < num2.length() ? num1.toCharArray() : num2.toCharArray();
    boolean carry = false;
    int i = right.length - 1;
    int j = left.length - 1;
    for (; i >= 0; i--, j--) carry = addStrings(sum, left[j], right[i], carry);
    for (; j >= 0; j--) carry = addStrings(sum, left[j], '0', carry);
    if (carry) sum.append(1);
    return sum.reverse().toString();
  }

  private static boolean addStrings(StringBuilder sumBuilder, char c, char d, boolean carry) {
    int sum = (c - '0') + (d - '0');
    sum += carry ? 1 : 0;
    sumBuilder.append(sum % 10);
    return sum >= 10;
  }
}
